vm/src/object/gc/trace.rs

```diff
@@ -0,0 +1,99 @@
+use rustpython_common::lock::PyMutex;
+
+use crate::object::gc::header::GcHeader;
+use crate::object::PyObjectPayload;
+use crate::{PyObjectRef, PyRef, AsObject};
+use core::ptr::NonNull;
+
+/// indicate what to do with the object afer calling dec()
+#[derive(PartialEq, Eq)]
+pub enum GcStatus {
+    /// should be drop by caller
+    ShouldDrop,
+    /// already buffered, will be drop by collector, no more action is required at caller
+    Buffered,
+    /// should keep and not drop by caller
+    ShouldKeep,
+}
+pub trait GcObjPtr: GcTrace {
+    fn inc(&self);
+    fn dec(&self) -> GcStatus;
+    fn rc(&self) -> usize;
+    /// return object header
+    fn header(&self) -> &GcHeader;
+    // as a NonNull pointer to a gc managed object
+    fn as_ptr(&self) -> NonNull<dyn GcObjPtr>;
+}
+
+/// use `trace()` to call on all owned ObjectRef
+pub trait GcTrace {
+    /// call tracer_fn for every GcOjbect owned by a dyn GcTrace Object
+    /// # API Contract
+    /// must make sure that every owned object(Every stored `PyObjectRef` to be exactly) is called with tracer_fn at most once.
+    /// 
+    /// if some field is not called, the worse results is memory leak, but if some field is called repeatly, panic and deadlock can happen.
+    /// 
+    /// Note that Two `PyObjectRef` to the Same `PyObject` still count as two Ref, and should be called twice(once for each one) in this case.
+    ///
+    /// ```
+    /// for ch in childs:
+    ///     tracer_fn(ch)
+    /// ```
+    /// *DO NOT* clone a `PyObjectRef`, use `as_ptr()` instead and operate on NonNull
```

Comment:
Is there any reason for `PyObjectRef` to be `Clone`?

Reply:
clone PyObjectRef in trace() will mess up ref cnt cycle collector, but it is very tempting to do so, like in PySet.keys() give vec of PyObjectRef(which invoke clone())

> Is there any reason for `PyObjectRef` to be `Clone`?

Reply:
In `vm/src/object/core.rs`, have you tried deleting `PyObjectRef`'s implementation of the `Clone` trait (or gating it behind the _absence_ of the `gc` feature)? If so, what consequences are there?

https://github.com/RustPython/RustPython/blob/c8c9c7b5f9815b376c584177b6b91a779026e4c3/vm/src/object/core.rs#L610-L615

Reply:
> In `vm/src/object/core.rs`, have you tried deleting `PyObjectRef`'s implementation of the `Clone` trait (or gating it behind the _absence_ of the `gc` feature)? If so, what consequences are there?
> 
> https://github.com/RustPython/RustPython/blob/c8c9c7b5f9815b376c584177b6b91a779026e4c3/vm/src/object/core.rs#L610-L615

unfortunatly, I have to rely on `PyObjectRef`'s clone() in everywhere else(except in trace()) to make the ref count system work properly, so no, I can't delete Clone for PyObjectRef, but I think we can make trace unsafe so people take those code seriously?